### ISSUE TITLE
Starting from 3.19.0 the code generated by Kotlin codegen (with custom strategy) no longer compiles

### ISSUE THREAD

**Issue by cies** [jOOQ/jOOQ#17374]
### Expected behavior

Should compile like it did in the 3.18.x, 3.17.x, 3.16.x, ... I forgot how long we use jOOQ now. But it always "just compiled".

### Actual behavior

Many compile errors like this:

```
e: build/generated-src/jooq/main/org/jooq/generated/tables/records/SeriesRecord.kt:21:83 Argument type mismatch: actual type is 'org.jooq.generated.tables.SeriesTbl.SeriesTbl', but 'org.jooq.Table<R!>!' was expected.
e: build/generated-src/jooq/main/org/jooq/generated/tables/records/SeriesRecord.kt:51:60 Unresolved reference 'ID'.
e: build/generated-src/jooq/main/org/jooq/generated/tables/records/SeriesRecord.kt:52:69 Unresolved reference 'CREATED_AT'.
e: build/generated-src/jooq/main/org/jooq/generated/tables/records/SeriesRecord.kt:53:69 Unresolved reference 'MODIFIED_AT'.
e: build/generated-src/jooq/main/org/jooq/generated/tables/records/SeriesRecord.kt:54:60 Unresolved reference 'ORGANIZATION_ID'.
e: build/generated-src/jooq/main/org/jooq/generated/tables/records/SignInTokenRecord.kt:21:93 Argument type mismatch: actual type is 'org.jooq.generated.table
```

And this:

```
e: build/generated-src/jooq/main/org/jooq/generated/tables/references/Tables.kt:287:27 Initializer type mismatch: expected 'org.jooq.generated.tables.AccountTbl', actual 'org.jooq.generated.tables.AccountTbl.AccountTbl'.
e: build/generated-src/jooq/main/org/jooq/generated/tables/references/Tables.kt:292:27 Initializer type mismatch: expected 'org.jooq.generated.tables.AddressTbl', actual 'org.jooq.generated.tables.AddressTbl.AddressTbl'.
e: build/generated-src/jooq/main/org/jooq/generated/tables/references/Tables.kt:297:36 Initializer type mismatch: expected 'org.jooq.generated.tables.AdvanceFormTbl', actual 'org.jooq.generated.tables.AdvanceFormTbl.AdvanceFormTbl'.
```

### Steps to reproduce the problem

```
package nl.stager.jooq

import java.util.Locale
import java.util.regex.Pattern
import org.jooq.SQLDialect
import org.jooq.codegen.DefaultGeneratorStrategy
import org.jooq.codegen.GeneratorStrategy
import org.jooq.meta.CatalogDefinition
import org.jooq.meta.Definition
import org.jooq.meta.ForeignKeyDefinition
import org.jooq.meta.IdentityDefinition
import org.jooq.meta.IndexDefinition
import org.jooq.meta.SchemaDefinition

private val idRegex = "_id".toRegex()
private val underscoreRegex = "_".toRegex()
private val columnNameRegex = Pattern.compile("(?<=[a-z])[A-Z]")

  /** Our custom generator strategy for name conversions. */
@Suppress("unused") // Used by the jOOQ Gradle plugin

class StagerGeneratorStrategy : DefaultGeneratorStrategy() {

  /** Despite the name this is also used when generating Kotlin code. */
  override fun getJavaClassName(definition: Definition, mode: GeneratorStrategy.Mode): String =
    composeClassName(definition, mode) // Our table names need little changes to be class names

  /** Despite the name this is also used when generating Kotlin code. */
  override fun getJavaIdentifier(definition: Definition): String {
    // [#1473] Identity identifiers should not be renamed by custom strategies
    if (definition is IdentityDefinition)
      return "IDENTITY_${getJavaIdentifier(definition.column.container)}"
    else if (definition is CatalogDefinition && definition.isDefaultCatalog)
      return "DEFAULT_CATALOG"
    else if (definition is SchemaDefinition && definition.isDefaultSchema)
      return "DEFAULT_SCHEMA"
    else if (definition is IndexDefinition
      && listOf(SQLDialect.MYSQL, SQLDialect.MARIADB)
        .contains(definition.getDatabase().dialect.family())
    ) {
      val tablePart = definition.table.outputName.uppercase(Locale.getDefault())
      val indexPart = definition.getOutputName().uppercase(Locale.getDefault())
      return "${tablePart}_$indexPart"
    } else if (definition is ForeignKeyDefinition) {
      val tablePart = definition.table.outputName.uppercase(Locale.getDefault())
      val columnsPart =
        definition.keyColumns.joinToString("__") { it.outputName.uppercase(Locale.getDefault()) }
      return "FK_${tablePart}__$columnsPart"
    }

    // https://github.com/jOOQ/jOOQ/blob/93ece30335883779a0519086be735a94eaced022/jOOQ-codegen/src/main/java/org/jooq/codegen/DefaultGeneratorStrategy.java#L132
    // Make our ugly "*_id" column name bits pretty.
    var result = definition.outputName.replace(idRegex, "Id")
    result = result.replace(underscoreRegex, "__")
    val m = columnNameRegex.matcher(result)
    val sb = StringBuilder()
    while (m.find()) {
      m.appendReplacement(sb, "_" + m.group().lowercase(Locale.getDefault()))
    }
    m.appendTail(sb)
    return sb.toString().uppercase(Locale.getDefault())
  }

  /** Despite the name this is also used when generating Kotlin code. */
  override fun getJavaMemberName(definition: Definition, mode: GeneratorStrategy.Mode): String =
    underscoreIdToCamelCaseId(definition)

  private fun adjustGetterSetterSuffix(definition: Definition): String =
    capitalize(underscoreIdToCamelCaseId(definition))

  private fun underscoreIdToCamelCaseId(definition: Definition): String =
    definition.outputName.replace(idRegex, "Id")

  companion object {
    private fun composeClassName(definition: Definition, mode: GeneratorStrategy.Mode): String {
      // NOTE(cies): adapted from code found here
      // https://github.com/jOOQ/jOOQ/blob/main/jOOQ-codegen/src/main/java/org/jooq/codegen/DefaultGeneratorStrategy.java
      val result = StringBuilder()
      // [#4562] Some characters should be treated like underscore
      result.append(
        capitalize(
          definition
            .outputName
            .replace(' ', '_')
            .replace('-', '_')
            .replace('.', '_')
            .replace(underscoreRegex, "To") // replace underscores (used in join tables) with "To"
        )
      )
      when (mode) {
        GeneratorStrategy.Mode.RECORD -> result.append("Record")
        GeneratorStrategy.Mode.DAO -> result.append("Dao")
        GeneratorStrategy.Mode.INTERFACE -> result.insert(0, "I")
        GeneratorStrategy.Mode.POJO -> result.append("Pojo")
        else -> result.append("Tbl") // to differentiate with Hibernate class names
      }
      return result.toString()
    }

    private fun capitalize(str: String): String {
      return if (str.isBlank()) str else str.substring(0, 1)
        .uppercase(Locale.getDefault()) + str.substring(1)
    }
  }
}
```

```
package nl.stager.jooq

import java.util.Locale
import org.jooq.codegen.GeneratorStrategy
import org.jooq.codegen.JavaWriter
import org.jooq.codegen.KotlinGenerator
import org.jooq.meta.Definition
import org.jooq.meta.TableDefinition

@Suppress("unused") // Referred to and called by the jooq generate Gradle plugin
class KotlinGeneratorWithFilters : KotlinGenerator() {
  internal enum class FilterableColumn {
    ORGANIZATION_ID, DELETED
  }

  override fun generateTableClassFooter(table: TableDefinition, out: JavaWriter) {
    val columnNames = table.columns.map { it.name }
    val matchingColumns = FilterableColumn.entries
      .filter { columnNames.contains(it.name.lowercase(Locale.getDefault())) }
    if (matchingColumns.isNotEmpty()) {
      out.println()
      out.println("// -------------------------------------------------------------------------")
      out.println("// THE CODE BELOW IS GENERATED BY STAGER (%s)", this.javaClass.simpleName)
      out.println("// -------------------------------------------------------------------------")
    }
    if (matchingColumns.contains(FilterableColumn.DELETED)) {
      out.println()
      generateNotDeletedMethod(out)
    }
    if (matchingColumns.contains(FilterableColumn.ORGANIZATION_ID)) {
      out.println()
      generateOrganizationMethod(out)
    }
    if (matchingColumns.size > 1) {
      out.println()
      generateAllFiltersMethod(out, matchingColumns)
    }
  }

  /** We add USELESS_CAST as this is sometimes the case and DEPRECATION for the COORDINATES type */
  override fun printClassAnnotations(
    out: JavaWriter,
    definition: Definition,
    mode: GeneratorStrategy.Mode
  ) {
    out.println("@Suppress(\"UNCHECKED_CAST\", \"USELESS_CAST\", \"DEPRECATION\")")
  }

  private fun generateAllFiltersMethod(out: JavaWriter, matchingColumns: List<FilterableColumn>) {
    val args = generateAllFiltersMethodParameters(matchingColumns)
    out.println("fun filterOrganizationAndNotDeleted($args): org.jooq.Condition {")
    out.println("return ${generateAllFiltersCondition(matchingColumns)}")
    out.println("}")
  }

  private fun generateAllFiltersMethodParameters(matchingColumns: List<FilterableColumn>): String {
    return matchingColumns
      .filter { it == FilterableColumn.ORGANIZATION_ID }
      .joinToString(", ") { generateOrganizationMethodParameter() }
  }

  private fun generateAllFiltersCondition(matchingColumns: List<FilterableColumn>): String {
    val filterConditions = matchingColumns.map {
      when (it) {
        FilterableColumn.DELETED -> generateNotDeletedCondition()
        FilterableColumn.ORGANIZATION_ID -> generateOrganizationCondition()
      }
    }
    val sb = StringBuilder()
    for (i in filterConditions.indices) {
      val condition = filterConditions[i]
      if (i == 0) {
        sb.append(condition)
      } else {
        sb.append(".and(").append(condition).append(")")
      }
    }
    return sb.toString()
  }

  private fun generateNotDeletedMethod(out: JavaWriter) {
    out.println("fun filterNotDeleted(): org.jooq.Condition {")
    out.println("return ${generateNotDeletedCondition()}")
    out.println("}")
  }

  private fun generateNotDeletedCondition(): String = "this.DELETED.isFalse"

  private fun generateOrganizationMethod(out: JavaWriter) {
    val organizationArg = generateOrganizationMethodParameter()
    out.println("fun filterOrganization($organizationArg): org.jooq.Condition {")
    out.println("return ${generateOrganizationCondition()}")
    out.println("}")
  }

  private fun generateOrganizationMethodParameter(): String = "organizationId: Long"

  private fun generateOrganizationCondition(): String = "this.ORGANIZATION_ID.eq(organizationId)"
}
```

```
jooq { // `generateJooq`
  // Gradle plugin should use the same version as the application, so we get adequate config options
  version.set(libs.versions.jooqVersion)
  edition.set(JooqEdition.OSS)
  configurations {
    create("main") {
      jooqConfiguration.apply {
        logging = org.jooq.meta.jaxb.Logging.WARN
        jdbc.apply {
          url = jdbcUrl
          user = dbUser
          password = dbPassword
        }
        generator.apply {
          // Subclass of `KotlinGenerator` with our table-aware filter methods
          name = "nl.stager.jooq.KotlinGeneratorWithFilters"
          strategy.apply {
            name = "nl.stager.jooq.StagerGeneratorStrategy" // use the custom generator strategy
          }
          generate.apply {
            withRelations(true) // Primary key / foreign key relations should be generated and used
            withRecords(true) // Generate jOOQ Record classes for type-safe querying
            withJavaTimeTypes(true) // Use `java.time.*` (JSR-310) instead of `java.sql.*` types

            withDeprecated(false) // Generate deprecated code for backwards compatibility
            withPojos(false) // No POJOs, as Records (that carry a db connection) are not problem for us
            withImmutablePojos(false) // Do not generate immutable POJOs
            withInterfaces(false) // Do not generate interfaces (implemented by records/POJOs)
            withDaos(false) // No DAOs, we make our own XyzRepository classes/objects
            withJpaAnnotations(false) // JPA annotations could help mapping Records to POJOs or Entities
            withValidationAnnotations(false) // Some Bean related annotations (trying to use less Bean spec)
            withFluentSetters(true) // Fluent setters are against the Bean spec (I'm also against the Bean spec)

            // These add/remove nullability question marks (`?`) to generated Kotlin code
            withKotlinNotNullRecordAttributes(true) // We do generate Record classes
            withKotlinNotNullPojoAttributes(true) // We do NOT generate Pojo classes
            withKotlinNotNullInterfaceAttributes(true) // We do NOT generate interfaces

            // We no longer need Java interop and the @Suppress("INAPPLICABLE_JVM_NAME") annotation
            // resulted in warnings with Kotlin 2.0+.
            withKotlinSetterJvmNameAnnotationsOnIsPrefix(false)

            // #### NOTE from Lukas Eder (Mr Jooq) on "nullability in SQL DSLs"
            // While it is tempting to use non-nullability guarantees that are derived from your
            // table definitions in generated code, it is important to remember that in SQL,
            // there are numerous reasons why an expression that is based on a column declared as
            // SQL NOT NULL can become NULL through the use of SQL operators, such as LEFT JOIN or
            // UNION, etc. In particular, a column with a DEFAULT or IDENTITY expression is
            // "nullable on write", so jOOQ will generate it as nullable despite a NOT NULL constraint!
          }

          database.apply {
            name = "org.jooq.meta.mysql.MySQLDatabase"
            inputSchema = "stager"
            withOutputSchemaToDefault(true) // removes all db name prefixing

            // Include all tables (allow first)
            includes = ".*"
            // Exclude the table that is used when running migrations (deny second)
            excludes = "flyway_schema_history"

            // `Account.active` not exposed by jOOQ: use `Contact.deleted` or `User.deleted`.
            // In similar vein `AccountEntity` has no `active` field in it's Hibernate model.
            // We have a NULLABLE `active` column on the `Account` table to allow a unique index
            // on `Account(organization_id, email, active)`, which works because unique indices
            // do not consider NULL values to be the same!
            // This is the default behavior on MySQL and Postgres (PG15 has `NULLS NOT DISTINCT`).
            // The `Account.active` mirrors the `Contact.deleted` column with a trigger, this
            // because the `deleted` column is required by SoftDeletableEntity and the trashcan.
            withIncludeExcludeColumns(true)
            withExcludes("Account.active")

            // A convenience map to create ForcedTypes for enums. Key is the db's `table.field`, value references the enum.
            forcedTypes.addAll(
              mapOf(
                // Organization
                "OrganizationDetails.kycStatus" to "models.organization.OrganizationDetailsEntity.KycStatus",
                "OrganizationDetails.customerType" to "models.organization.OrganizationDetailsEntity.CustomerType",
                "OrganizationSettings.addressRequirement" to "models.organization.OrganizationSettingsEntity.Optionality",
                "OrganizationSettings.zipCodeRequirement" to "models.organization.OrganizationSettingsEntity.Optionality",
// ... removed many from the map
              ).map {
                ForcedType()
                  .withEnumConverter(true)
                  .withIncludeExpression(".*\\.${it.key.replace(".", "\\.")}")
                  .withUserType(it.value)
              }
            )

            forcedTypes.addAll(
              arrayOf(
                // Forces the TINYINT values to a boolean expression in the table fields for jOOQ
                // https://blog.jooq.org/how-to-map-mysqls-tinyint1-to-boolean-in-jooq/

                // In the future we want to get rid of this forced type. We want to move the
                // boolean types to the MySQL `BIT(1)` type. jOOQ maps this to a boolean by default
                // which would make this obsolete
                ForcedType().withName("BOOLEAN").withIncludeTypes("(?i:TINYINT\\(1\\))")
              )
            )
          }
        }
      }
      // Not specifying "target" puts generated source in: `build/generated-src/jooq/`
    }
  }
}
```

### jOOQ Version

3.19.x

### Database product and version

MySQL

### Java Version

JDK21, Kotlin 2.0.20

### JDBC / R2DBC driver name and version (include name if unofficial driver)

{ group = "com.mysql", name = "mysql-connector-j", version = "9.0.0" } # also happened with 8.0.33

**Comment by lukaseder** [comment]
Thanks for your report.

Ideally, an MCVE based on our template would help speed up analysis: https://github.com/jOOQ/jOOQ-mcve, the focus being on minimal here (to avoid investigating things that you shared but aren't really relevant).

If that's too much work, can you please post some example classes that don't compile

**Comment by cies** [comment]
I'm also digging deeper into it today. I just wanted to raise the issue first -- knowing that you reply quick -- to know for sure im not spending time of an issue that's already known.

Also, I expect it may be because the our own generation strategy lacks in some aspect.

Later today I'll report my findings here.

**Comment by lukaseder** [comment]
> to know for sure im not spending time of an issue that's already known.

This kind of significant regression is usually fixed very soon after it is known.

> Also, I expect it may be because the our own generation strategy lacks in some aspect.

My assumption is more that you're doing something that exposes an actual flaw in jOOQ's usage of the strategy. If I could see some generation output, that would really help.

**Comment by cies** [comment]
Here you go. https://gist.github.com/cies/5d5d24df1497acfb78e7faf4520e259f

I had a look also at making a MCVE, but with our generator is was going to be more work than i had time for today. And this was rather quick, I already had them files for my own research.

Around these lines I found something off (in the 3.19.0 file obviously):

```kotlin
import org.jooq.generated.indexes.ADDRESS_IX_ADDRESS__COUNTRY_ID
import org.jooq.generated.keys.FK_ADDRESS__COUNTRY_ID
import org.jooq.generated.keys.FK__BILLING_INFORMATION____REGISTERED_ADDRESS_ID
import org.jooq.generated.keys.FK__BULK_CONTACT_EDIT____VISIT_ADDRESS_ID
import org.jooq.generated.keys.FK__CONTACT____POST_ADDRESS_ID
import org.jooq.generated.keys.FK__CONTACT____VISIT_ADDRESS_ID
import org.jooq.generated.keys.FK__ORDER_RECEIPT____ADDRESS_ID
import org.jooq.generated.keys.KEY__ADDRESS__PRIMARY
import org.jooq.generated.tables.BillingInformationTbl.BillingInformationTbl
import org.jooq.generated.tables.BulkContactEditTbl.BulkContactEditTbl
import org.jooq.generated.tables.ContactTbl.ContactTbl
import org.jooq.generated.tables.CountryTbl.CountryTbl
import org.jooq.generated.tables.OrderReceiptTbl.OrderReceiptTbl
```

The number of "FK" imports became more compared to 3.18.6, and the "tables" imports that have the Class name (`*Tbl`) doubled also seems not okay.

The double `*Tbl` thing I can link to the second error snippet i started my issue with, i.e.:

```
e: build/generated-src/jooq/main/org/jooq/generated/tables/references/Tables.kt:292:27 Initializer type mismatch: expected 'org.jooq.generated.tables.AddressTbl', actual 'org.jooq.generated.tables.AddressTbl.AddressTbl'.
```

Thanks for looking into it.

**Comment by lukaseder** [comment]
Thanks for your update. I appreciate you trying to investigate the cause of this issue on your end, but if you want *me* to help, can you please do post the classes that have an actual compilation error? E.g. these ones:

```
build/generated-src/jooq/main/org/jooq/generated/tables/references/Tables.kt
build/generated-src/jooq/main/org/jooq/generated/tables/records/SeriesRecord.kt
```

The error is:

> Initializer type mismatch: expected 'org.jooq.generated.tables.AddressTbl', actual 'org.jooq.generated.tables.AddressTbl.AddressTbl'.

So, how can I help you investigate *this particular* error? Making an MCVE may be a lot of work indeed, but just posting these classes with compilation errors would really be helpful, no?

The compilation error is in the reference to the table identifiers, not in the table definition itself.

> The number of "FK" imports became more compared to 3.18.6, 

You can see that the `FK` references are necessary to the to-many relationship paths, which is a new jOOQ 3.19 feature. You can turn off the feature and the new imports go away, but I'm not sure how that's related to any potential issue here.

> and the "tables" imports that have the Class name (`*Tbl`) doubled also seems not okay.

They're required for the to-many relationship paths again. E.g. look at `fkContact_PostaddressId()`.

**Comment by cies** [comment]
> Thanks for your update. I appreciate you trying to investigate the cause of this issue on your end, but if you want _me_ to help, can you please do post the classes that have an actual compilation error? E.g. these ones:
> 
> ```
> build/generated-src/jooq/main/org/jooq/generated/tables/references/Tables.kt
> build/generated-src/jooq/main/org/jooq/generated/tables/records/SeriesRecord.kt
> ```

I've added them to the Gist: https://gist.github.com/cies/5d5d24df1497acfb78e7faf4520e259f

> The error is:
> 
> > Initializer type mismatch: expected 'org.jooq.generated.tables.AddressTbl', actual 'org.jooq.generated.tables.AddressTbl.AddressTbl'.

It's just one of the errors, the errors are on nearly every file so I just pasted an excerpt of the errors.

> So, how can I help you investigate _this particular_ error? Making an MCVE may be a lot of work indeed, but just posting these classes with compilation errors would really be helpful, no?

You are so right.

> You can see that the `FK` references are necessary to the to-many relationship paths, which is a new jOOQ 3.19 feature. You can turn off the feature and the new imports go away, but I'm not sure how that's related to any potential issue here.
> 
> They're required for the to-many relationship paths again. E.g. look at `fkContact_PostaddressId()`.

Clear.

**Comment by lukaseder** [comment]
> I've added them to the Gist: https://gist.github.com/cies/5d5d24df1497acfb78e7faf4520e259f

1. `[jooqgen-3.19.12-tables-records-SeriesRecord.kt](https://gist.github.com/cies/5d5d24df1497acfb78e7faf4520e259f#file-jooqgen-3-19-12-tables-records-seriesrecord-kt)` doesn't contain the correct content, I'm guessing that's a copy paste error
2. [jooqgen-3.19.12-tables-references-Tables.kt](https://gist.github.com/cies/5d5d24df1497acfb78e7faf4520e259f#file-jooqgen-3-19-12-tables-references-tables-kt) has been truncated at line 122, but in your original report, the error was at line 287, see: `build/generated-src/jooq/main/org/jooq/generated/tables/references/Tables.kt:287:27`

> It's just one of the errors, the errors are on nearly every file so I just pasted an excerpt of the errors.

I understand that many files have errors, but I'm afraid, you haven't pasted any errors as far as I can tell. You've pasted code generation output, but not the parts that show the errors.

**Comment by lukaseder** [comment]
OK, I have another hypothesis for you to check. A new thing that has been added to code generation output is path classes, e.g.:

```kotlin
    open class AddressTbl : AddressTbl, Path<AddressRecord> {
        constructor(path: Table<out Record>, childPath: ForeignKey<out Record, AddressRecord>?, parentPath: InverseForeignKey<out Record, AddressRecord>?): super(path, childPath, parentPath)
        private constructor(alias: Name, aliased: Table<AddressRecord>): super(alias, aliased)
        override fun `as`(alias: String): AddressTbl = AddressTbl(DSL.name(alias), this)
        override fun `as`(alias: Name): AddressTbl = AddressTbl(alias, this)
        override fun `as`(alias: Table<*>): AddressTbl = AddressTbl(alias.qualifiedName, this)
    }
```

The way you rename classes is wrong in my opinion. In `composeClassName()`, you default to suffixing everything "else" as `Tbl`:

```kotlin
      when (mode) {
        GeneratorStrategy.Mode.RECORD -> result.append("Record")
        GeneratorStrategy.Mode.DAO -> result.append("Dao")
        GeneratorStrategy.Mode.INTERFACE -> result.insert(0, "I")
        GeneratorStrategy.Mode.POJO -> result.append("Pojo")
        else -> result.append("Tbl") // to differentiate with Hibernate class names
      }
```

But that's not correct. Your switch here should be exhaustive. When new `Mode` values are added, you shouldn't just suffix them all as `Tbl`, e.g. `Mode.PATH`

Can you try adding this?

```kotlin
        GeneratorStrategy.Mode.PATH -> result.append("Path")
```

**Comment by lukaseder** [comment]
If that's it, then I guess my confusion here is that the kotlin compiler is getting confused by nested classes that share the same name as the enclosing class, especially when there's a companion object. Though, I'm surprised you wouldn't have seen any compilation error like this:

```kotlin
open class A {
    open class A : A {}
}
```

Here, I'm getting:

> Cycle formed in the inheritance hierarchy of this type.

**Comment by lukaseder** [comment]
I can confirm this is the problem here. Your strategy produces a nested class of the same name as the enclosing class, which is a problem in all languages, but the actual error is just one error message hidden in a ton of error messages.

The confusion in the discussion of this issue was mainly because the important error message wasn't shared as it was hidden in a ton of side-effect error messages. Instead the discussion focused only on the side effects (e.g. the referencing of table classes).

I've created an issue to address this:

- https://github.com/jOOQ/jOOQ/issues/17402

We'll append a `"Path"` suffix to nested `Path` classes if a user `GeneratorStrategy` fails to name it differently from the enclosing `Table` class.

**Comment by lukaseder** [comment]
#17402 is fixed for 3.20.0 and 3.19.14. Thanks again for your report and for your patience!

**Comment by cies** [comment]
@lukaseder Thank you for jOOQ :)

Based on your remarks we have:
* dropped the `Tbl` suffix (was not needed anymore since we now have `Entity` suffices for all Hibernate-managed classes)
* made case statements exhaustive and *without* a default (`else` in Koltin`) clause, so if jOOQ extends the enum we get the compile error we want
* added a handling of `DefaultInverseForeignKeyDefinition` in our overridden implementation of `getJavaIdentifier`

There's one problem that still remains: our generator generates some import statements on top of a class that extends `TableImpl`, that does not reference an existing symbol! This only happens for definitions of type `DefaultInverseForeignKeyDefinition`, which, in our case, alway come in the form of `FK_TABLENAME__ID`.  E.g.:

```kotlin
// from different files:
import org.jooq.generated.keys.FK_AUTOMATION__ID
import org.jooq.generated.keys.FK_ACCOUNT__ID
// ...
```

I will wait for the release of 3.19.14 (as we are not in a hurry at all - 3.18.x works great for ut) and try to see if I can debug our last issue when using that. In case I dont manage fixing it myself, I will open another ticket.
